components/eventing-controller/controllers/subscription/nats/integration_test.go

```diff
@@ -0,0 +1,837 @@
+			subscriptionOpts: []reconcilertesting.SubscriptionOpt{
+				reconcilertesting.WithFilter("", reconcilertesting.OrderCreatedEventTypeNotClean),
+				reconcilertesting.WithSinkURL(validSinkURL(ens, ":8080", "/myEndpoint")),
+			},
+			expect: expect{
+				k8sSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
+					reconcilertesting.HaveSubscriptionReady(),
+				},
+			},
+			shouldTestDeletion: true,
+		},
+
+		{
+			name: "empty protocol, protocol setting and dialect",
+			subscriptionOpts: []reconcilertesting.SubscriptionOpt{
+				reconcilertesting.WithFilter("", reconcilertesting.OrderCreatedEventTypeNotClean),
+				reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+			},
+			expect: expect{
+				k8sSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
+				},
+				natsSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.BeNotNil(),
+					reconcilertesting.BeValid(),
+					reconcilertesting.HaveSubject(reconcilertesting.OrderCreatedEventType),
+				},
+			},
+			shouldTestDeletion: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			subscription, subscriptionName := createSubscription(ctx, g, ens, tc.subscriptionOpts...)
+			testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName, tc.expect.k8sSubscription...)
+			testEventsOnK8s(ctx, g, ens, tc.expect.k8sEvents...)
+			testSubscriptionOnNATS(g, ens, subscriptionName, tc.expect.natsSubscription...)
+			testDeletionOnK8s(ctx, g, ens, subscription, tc.shouldTestDeletion)
+		})
+	}
+}
+
+//TestChangeSubscription tests if existing subscriptions are reconciled properly after getting changed.
+func TestChangeSubscription(t *testing.T) {
+	ctx := context.Background()
+	g := gomega.NewGomegaWithT(t)
+	ens := setupTestEnsemble(ctx, reconcilertesting.EventTypePrefix, g, 4224, 7074)
+	defer ens.cancel()
+
+	var testCases = []struct {
+		name               string
+		subscriptionOpts   []reconcilertesting.SubscriptionOpt
+		expectBefore       expect
+		changeSubscription func(subscription *eventingv1alpha1.Subscription)
+		expectAfter        expect
+		shouldTestDeletion bool
+	}{
+		{
+			name: "clean event types; add filters to subscription without filters",
+			subscriptionOpts: []reconcilertesting.SubscriptionOpt{
+				reconcilertesting.WithEmptyFilter(),
+				reconcilertesting.WithWebhookForNATS(),
+				reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+			},
+			expectBefore: expect{
+				k8sSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
+					reconcilertesting.HaveSubsConfiguration(configDefault(ens.defaultSubscriptionConfig.MaxInFlightMessages)),
+				},
+			},
+			changeSubscription: func(subscription *eventingv1alpha1.Subscription) {
+				eventTypes := []string{
+					uncleanEventType("0"),
```

Comment:
out-of-scope of your PR: IMHO the function would be more intuitive if it would be called `(make)UncleanOrderCreatedEventType("0")`. Same for the `cleanEventType ` function.

What is your opinion on this ?

Reply:
mhh, I guess I again created `uncleanEventType` with the idea of c#'s properties. I am also not completely happy with this.